common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java

```diff
@@ -0,0 +1,215 @@
+    public void init(Properties config) throws ServletException {
+        spnegoEnabled = Boolean.parseBoolean(config.getProperty(SPNEGO_ENABLED_CONFIG, Boolean.TRUE.toString()));
+        if (spnegoEnabled) {
+            super.init(config);
+        }
+
+        try {
+            provider = new KerberosAuthenticationProvider();
+            SunJaasKerberosClient client = new SunJaasKerberosClient();
+            client.setDebug(true);
+            provider.setKerberosClient(client);
+            provider.setUserDetailsService(new KerberosUserDetailsService());
+        } catch (Exception ex) {
+            LOG.error("Failed to initialize the Kerberos Login Authentication Handler.", ex);
+            throw new ServletException(ex);
+        }
+    }
+
+    @Override
+    public void destroy() {
+
+    }
+
+    /**
+     * Perform Basic authentication through Kerberos, if the authentication fails, delegates to KerberosAuthenticationHandler for SPNEGO exchange.
+     *
+     * @return an authentication token if the request is authorized or null
+     * @throws IOException
+     * @throws AuthenticationException
+     */
+    @Override
+    public AuthenticationToken authenticate(HttpServletRequest request, HttpServletResponse response) throws IOException, AuthenticationException {
+        AuthenticationToken token = kerberosLogin(request, response);
+
+        if (token == null && spnegoEnabled) {
+            LOG.debug("Attempting SPNEGO authentication sequence as kerberos login failed.");
+            token = super.authenticate(request, response);
+        }
+
+        return token;
+    }
+
+    /**
+     * Perform Basic Authentication using Kerberos credentials the http request.
+     * @return an AuthenticationToken on successful authentication or null.
+     * @throws IOException
+     */
+    private AuthenticationToken kerberosLogin(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        if (provider == null) {
+            LOG.error("The Kerberos authentication provider is not initialized.");
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return null;
+        }
+        String authorization =  request.getHeader(AUTHORIZATION_HEADER);
+        if (!request.getMethod().equals(METHOD) || !request.isSecure() || authorization.isEmpty()) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Kerberos Login is not attempted because method: {}, secure: {}, authorization is empty: {}", request.getMethod(),
+                        request.isSecure(), authorization.isEmpty());
+            }
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return null;
+        }
+        String credentials = authorization.split(BASIC_AUTHENTICATION)[1].trim();
+        byte[] credentialsArray = Base64.getDecoder().decode(credentials);
+        String[] userPassword = new String(credentialsArray, StandardCharsets.UTF_8).split(":");
```

Comment:
Can we rename this to "principalAndPassword"? I was bit confused when rawPrincipal was extracted from "userPassword".